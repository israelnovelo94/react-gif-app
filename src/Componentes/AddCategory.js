import React, { useState } from 'react'
import PropTypes from 'prop-types';

export const AddCategory = ({setCategories}) => {
  const [InputValue, setInputValue] = useState(''); //Enviamos string vacio para evitar el undefined y un error
  const handleInput = (e) => {
    setInputValue(e.target.value); //Actualizamos el valor del input
  }
  
  const handleSubmit = (e) => {
    e.preventDefault(); //prevenimos el refresh en la pagina
    if (InputValue.trim().length > 2) {
      setCategories(cats => [ InputValue, ...cats, ]);//Utilizamos la propiedad setCategories y añadimos el valor del input a la lista
      setInputValue('');
      
    }
  }
  return (
    <form onSubmit={handleSubmit}>
        <input 
          type="text"
          value={InputValue}
          onChange={handleInput}
        />
    </form>
  )
}
//Hacemos que la propiedad setCategorie sea obligatoria
AddCategory.propTypes = {
  setCategories: PropTypes.func.isRequired
}
