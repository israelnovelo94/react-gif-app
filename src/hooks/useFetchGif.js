import { useEffect, useState } from 'react';
import { getGift } from '../helpers/getGifs';

export const useFetchGif = (category) => {
    const [first, setfirst] = useState({
        data: [],
        loading: true
    });
    useEffect(() => {
        // setTimeout(() => {
            getGift(category).then( img => {
                setfirst({
                    data: img,
                    loading: false
                });
            }); 
        // }, 3000);
    }, [category]);
    return first;
}
