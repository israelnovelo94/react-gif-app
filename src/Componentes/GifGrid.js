import React from 'react'
import { useFetchGif } from '../hooks/useFetchGif';
import { GifGridItem } from './GifGridItem';

export const GifGrid = ({category}) => {

    // Llamamos al custom hook y hacemos la desestructuracion de los datos
    const {data,loading} = useFetchGif(category);

  return (
    <>
        <h3>{category}</h3>
        {/* Hacemos una validacion para desplegar el loading en lo que carga la información */}
        {loading && <h4>Loading...</h4>}
        <div className='card-grid'>
            
            { 
                //mapeamos o recorremos la data para sacar la informacion del arreglo
                data.map( img => 
                  
                    <GifGridItem 
                    key={img.id}
                    {...img}
                    />
                    ) 
            }
            

        </div> 
    </>
  )
}
