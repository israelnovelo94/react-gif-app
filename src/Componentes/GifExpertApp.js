import React, { useState } from 'react';
import { AddCategory } from './AddCategory';
import { GifGrid } from './GifGrid';

const GifExpertApp = ( {defaultCategories = []}) => {
    

    const [categories, setCategories] = useState(defaultCategories);

    // const handleAdd = () => {
    //     setCategories( cats => [...cats, 'HunterxHunter']);
    // }
    return (
        <>
        <h2>Gix Expert App</h2>
        <hr />

        <AddCategory setCategories={ setCategories } />
        <ol>
        {
            categories.map( (category) => 

                    <GifGrid 
                        key={ category }
                        category={ category }
                    />
            )
        }
        </ol>
        </>
        
        
    )
}

export default GifExpertApp;