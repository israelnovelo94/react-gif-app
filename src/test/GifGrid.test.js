import { shallow } from 'enzyme';
import '@testing-library/jest-dom';
import { GifGrid } from '../Componentes/GifGrid';
import { useFetchGif } from '../hooks/useFetchGif';
jest.mock('../hooks/useFetchGif');

describe('Probando el GifGrid', () => {


    test('snapshot del GifGrid', () => {
        useFetchGif.mockReturnValue({
            data: [],
            loading: true
        });
        const category = 'One Punch'
        const wrapper = shallow(<GifGrid category={category} />); 
        expect( wrapper ).toMatchSnapshot();
    });

    test('simulando el uso del useFetchGif', () => {
        const gifs = [{
            id: 1,
            url: "www.google.com",
            title: "TEST"
        }]
        useFetchGif.mockReturnValue({
            data: gifs,
            loading: false
        });
        const category = 'One Punch'
        const wrapper = shallow(<GifGrid category={category} />); 
        expect( wrapper ).toMatchSnapshot();

        expect( wrapper.find('p').exists() ).toBe(false);
        expect( wrapper.find('GifGridItem').length ).toBe(gifs.length);


    });

});