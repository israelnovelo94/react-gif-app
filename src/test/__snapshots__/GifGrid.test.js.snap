// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Probando el GifGrid simulando el uso del useFetchGif 1`] = `
<Fragment>
  <h3>
    One Punch
  </h3>
  <div
    className="card-grid"
  >
    <GifGridItem
      id={1}
      key="1"
      title="TEST"
      url="www.google.com"
    />
  </div>
</Fragment>
`;

exports[`Probando el GifGrid snapshot del GifGrid 1`] = `
<Fragment>
  <h3>
    One Punch
  </h3>
  <h4>
    Loading...
  </h4>
  <div
    className="card-grid"
  />
</Fragment>
`;
